enhancements/cli-in-cluster-management.md

```diff
@@ -0,0 +1,172 @@
+
+## Summary
+
+This proposal is describing the mechanism for how CLI authors can deliver tools to OpenShift clusters.  An in-cluster feature
+is needed to manage various CLIs available for OpenShift and related services.  
+The goal is for a user to discover, install and upgrade tools easily and from a single
+location.  
+
+Each component is responsible for building and publishing its artifacts and registering information regarding supplied binaries.
+Currently, that location is [index of /pub/openshift-v4/clients](https://mirror.openshift.com/pub/openshift-v4/clients/),
+and this makes it difficult for disconnected installations to mirror them.
+
+The CLI in-cluster manager will retrieve binaries from images that package a CLI's artifacts.  Through the use of local mirrored
+registries, disconnected environments will have access to OpenShift CLIs.  Currently, we provide disconnected environments 
+cli-artifacts and oc download links that do not require anything outside the cluster.  The goal is to provide the same for other
+binaries/tools/CLIs.
+
+The CLI in-cluster manager will provide a mechanism to provide, list, install, and upgrade OpenShift tools/plugins/CLIs from
+within a cluster.  The [ConsoleCLIDownload CRD](https://github.com/openshift/api/blob/master/console/v1/0000_10_consoleclidownload.crd.yaml) 
+will be extended to meet the requirements of a new CLI manager.  Each CLI will provide necessary information about its provided artifacts,
+including its description and location within the image.  
+
+An oc subcommand, `oc krew`,  will allow users to interact with CLI CRs to list, update, and install CLIs in the $HOME directory.
+
+## Motivation
+
+As more services are created on top of OpenShift, more CLIs are introduced to simplify interaction with these services.
+Some current examples are oc, kubectl, odo, istio, tekton, and knative.  It is difficult for users to discover what tools exist, 
+where to download them from and which version they should download.  We need to simplify as much as possible the interaction
+of services on OpenShift.  We need a mechanism for providing and consuming tools that is simple to add on to as new CLIs are 
+developed from a variety of sources - and this needs to be managed from within a cluster and with disconnected installs.
+This is the motivation for the CLI in-cluster manager.
+
+### Goals
+
+Each component wishing to provide customers with their binaries in-cluster will build and publish artifacts via an official channel to a central index
+from which the CLI in-cluster manager can access them. 
+
+Possible routes for supplying CLI binaries:    
+1.  Central index that has links to where artifacts are stored.  This is what we have now with ConsoleCLIDownloads CRD. 
+The [ConsoleCLIDownload CRD](https://github.com/openshift/api/blob/master/console/v1/0000_10_consoleclidownload.crd.yaml) could be extended and reused.
```

Comment:
Drop this sentence of reusing console CRD too.

Reply:
dropped